### ISSUE TITLE
Handing of SetPropertyFromCommand results

### ISSUE THREAD

**Issue by Try2Code** [buildbot/buildbot#6303]
hi!

I want to let builders create a list tasks and create corresponding steps from it. For that I use 
```python
factory.addStep(steps.SetPropertyFromCommand(command="./config/buildbot/listExperiments",
  property="EXP_LIST")
explist = util.Interpolate('%(prop:EXP_LIST)s')
```
The content of the variable EXP_LIST on the worker is a simple string. Now I want to perform string operations on that string (like `split()`), but it does not work on explist directly. I know I have to perform some _rendering_ or _transformation_, but I cannot figure out from the documentation how to do it correctly.

In a `ShellCommand()` it can be used _as if_ it would be a string. I am also happy to write a class to perform operations on `util.Interpolate()` results, but currently I have no clue how to start with that. 

As a side note: Why is extra rendering even needed? The worker clearly has the correct value, The server lets the worker execute the corresponding command, utils.Interpolate() is told to interpret the property as a string, but still the result is something different. 

Don't get me wrong - this is not a complain or so. I use buildbot in my main project for about 10 years now and it saved our distributed group of developers from falling apart and made us a lot more productive. At the moment I concentrate on a major update of the server and would like to understand the 3.x version better in order to use it properly.

thx in advance

**Comment by Try2Code** [comment]
Or let's rephrase it: 
Is there **any** way to create steps dynamically by the worker based on let's say the compile-procedure done by a specific builder? 
I was reading through the docu about RemoteCommand, Master-Worker-API or even transferring files but nothing works so far ...

**Comment by vibbo** [comment]
Hi, just another Buildbot user here.

I am not sure if I fully understand your issue, so correct me if I am wrong. But I think you want to achieve the following:
* Execute first step on worker and obtain your tasks as string into property EXP_LIST.
* Based on parsing those tasks from property EXP_LIST, you want to execute additional steps (that those tasks represent)?

If so, I believe the easiest way is to have all those steps (tasks) already configured in factory and condition executing them during build, based on whether the task is in property EXP_LIST during that particular build.
So it could look something like this (not tested):
```python
from buildbot.plugins import steps, util
@util.renderer
def r_do_step(props, task):
    return task in props.getProperty("EXP_LIST").split()

...

factory.addStep(steps.SetPropertyFromCommand(command="./config/buildbot/listExperiments", property="EXP_LIST"))
factory.addStep(steps.ShellCommand(command=["make"], doStepIf=r_do_step.withArgs("make")))
factory.addStep(steps.ShellCommand(command=["test"], doStepIf=r_do_step.withArgs("test")))
factory.addStep(steps.ShellCommand(command=["package"], doStepIf=r_do_step.withArgs("package")))
```

**Comment by Try2Code** [comment]
hi - thx a lot for taking your time to look into it!

With your setup the different tests will be executed and logged into the same stdio stream. In my software I have 20-30 tests, each with 1k-50k output. In a single stdio stream there is no chance to keep the overview. moreover these tests run in parallel.

My current solution is to use the 'logfiles' argument of the corresponding shell commend and add each of the 20-30 logfiles. This way I can keep logs separate. But: If runs fail it's hard to find out, which one failed.
That's why I would like to use real steps running in parallel - here I get a nice indicator (red/green) per test. From what I understood reading the documentation, Triggerable steps can run in parallel, but again: What these steps should do is decided on the worker at runtime. So I cannot pre-define the steps at start time of the buildbot server. 

Maybe I am expecting the wrong thing from the buildbot framework - no clue if it is supposed to provide this kind of workflow...

**Comment by Try2Code** [comment]
Do you now how to correctly overwrite/extend the ShellCommand.run() methods. At the end of my new run() methods I should call the one from the master class, but I don't know how to do it

have good weekend!

**Comment by vibbo** [comment]
It's much more clear to me now, what you want to do.

**My preferred method**
If you have any control over your test output(s), you can generate a summary to stdout stream. It can even be color-coded to highlight failing/passing tests. Then you will know which test failed and examine the correct logfile.

This is sort of how I do it. I have a python executor script that runs my tests and it generates a summary. Executed and managed by single step.

Also similar use-case is described in https://docs.buildbot.net/current/manual/configuration/steps/shell_command.html?highlight=logfiles#buildbot.steps.shell.ShellCommand for parameter _logfiles_. 

**Dynamic steps** Note: I have not used this (yet :))
If you want to dynamically generate steps (per test) during the build, you probably can. But it would be a bit complicated. Moreover **they would be executed in sequence** - as steps are.

This is how steps are added to `Build` (`Build` instance is created for every new build by your preconfigured factory `BuildFactory`):
https://github.com/buildbot/buildbot/blob/dfd3cda7267d4e9d8bde61c3d4e5e2bec9b04d11/master/buildbot/process/build.py#L488-L498
So you could inject additional steps to running build in your own customized step. It would access the current build via `self.build`. But this can be tricky to implement correctly.

**Triggerable schedulers**
There are no triggerable steps I know of, only [Triggerable schedulers](https://docs.buildbot.net/current/manual/configuration/schedulers.html?highlight=triggerable#triggerable-scheduler). They can be triggered by [Trigger](https://docs.buildbot.net/current/manual/configuration/steps/trigger.html?highlight=triggerable#buildbot.steps.trigger.Trigger) step.

These are just builds that run in parallel during your trigger step. Much like [Buildbot's own build](https://buildbot.buildbot.net/#/builders/1/builds/2356) run multiple sub-build.

The hard part here is, that they run on their own worker(s) (sandbox) so you have to pass the context somehow. It is relatively easy, when you only tell all those triggered builds to compile certain revision from certain repository in specific configuration. This is done by passing these information to the triggered schedulers in properties. But a bit more complicated if you have build artefacts (in your case your tests) from your previous step and want to execute those in sub-plans.

**To you second comment** about overwriting/extending existing steps, I can recommend to have a look at https://github.com/buildbot/buildbot/tree/master/master/buildbot/steps. There you can find all sorts of more complex steps based on simple basic steps and mixins.

**Comment by Try2Code** [comment]
> It's much more clear to me now, what you want to do.
> 
> **My preferred method** If you have any control over your test output(s), you can generate a summary to stdout stream. It can even be color-coded to highlight failing/passing tests. Then you will know which test failed and examine the correct logfile.
> 
> This is sort of how I do it. I have a python executor script that runs my tests and it generates a summary. Executed and managed by single step.
> 

Yes, I also generate a summary for the tests. Otherwise developers would be lost in the logs


> Also similar use-case is described in https://docs.buildbot.net/current/manual/configuration/steps/shell_command.html?highlight=logfiles#buildbot.steps.shell.ShellCommand for parameter _logfiles_.
> 

That's what I currently use: submit the jobscripts into the queuing system with a certain logfile naming convention and let buildbot display each of them in an individual web box. What I want is to have each experiment run as a step so that I can skip the summary since the return value of each jobscript determines its success.

> **Dynamic steps** Note: I have not used this (yet :)) If you want to dynamically generate steps (per test) during the build, you probably can. But it would be a bit complicated. Moreover **they would be executed in sequence** - as steps are.

yh, I thought so. nothing about parallel step execution in the docu. 

> 
> This is how steps are added to `Build` (`Build` instance is created for every new build by your preconfigured factory `BuildFactory`):
> 
> https://github.com/buildbot/buildbot/blob/dfd3cda7267d4e9d8bde61c3d4e5e2bec9b04d11/master/buildbot/process/build.py#L488-L498
> 
> So you could inject additional steps to running build in your own customized step. It would access the current build via `self.build`. But this can be tricky to implement correctly.

Indeed. I have a hard time to figure our, when rendering is needed and when not

> 
> **Triggerable schedulers** There are no triggerable steps I know of, only [Triggerable schedulers](https://docs.buildbot.net/current/manual/configuration/schedulers.html?highlight=triggerable#triggerable-scheduler). They can be triggered by [Trigger](https://docs.buildbot.net/current/manual/configuration/steps/trigger.html?highlight=triggerable#buildbot.steps.trigger.Trigger) step.
> 
> These are just builds that run in parallel during your trigger step. Much like [Buildbot's own build](https://buildbot.buildbot.net/#/builders/1/builds/2356) run multiple sub-builds.
> 
> The hard part here is, that they run on their own worker(s) (sandbox) so you have to pass the context somehow. It is relatively easy, when you only tell all those triggered builds to compile certain revision from certain repository in specific configuration. This is done by passing these information to the triggered schedulers in properties. But a bit more complicated if you have build artefacts (in your case your tests) from your previous step and want to execute those in sub-plans.

hmm, .. this sounds way to complex for my rather simple task tbh. I guess the Triggerables are not suitable for what I want

> 
> **To you second comment** about overwriting/extending existing steps, I can recommend to have a look at https://github.com/buildbot/buildbot/tree/master/master/buildbot/steps. There you can find all sorts of more complex steps based on simple basic steps and mixins.

Thx. I look into the source code made it pretty clear:

```python
class MyCommandHandler(steps.ShellCommand):
  def __init__(self, arg, **kwargs):
    self.experiment_list = arg
    super().__init__(**kwargs)
  
 @defer.inlineCallbacks
  def run(self):
        # .... some other stuff with logfiles
         cmd = yield self.makeRemoteShellCommand()
         yield self.runCommand(cmd)
         return cmd.results()

```

With these 3 lines the command is correctly handled. arg is given as `util.Interpolate('%(prop:EXP_LIST)s')` in the constructor. I think I will stick to my logfile-solution. My users are used to it and the parallelism is mandatory for me.

Again: thx for your support! Sometimes I need human feedback - the docu of the corresponding parts boils down to a list of examples, which is not always helpful. 

Next item on my list: gitlab integration or _How to start buildbot by a PR comment?_

cheers :1st_place_medal:
